### PR TITLE
Parallel in-place matrix transposition and scaling

### PR DESCRIPTION
**Pull request by chillenb**
This PR adds a tiled parallel in-place matrix scaling / transposition routine to libnp_helper.
Read more about the algorithm at [this link](https://colfaxresearch.com/multithreaded-transposition-of-square-matrices-with-common-code-for-intel-xeon-processors-and-intel-xeon-phi-coprocessors/). I've added it to the 2D square inplace and (0,2,1) inplace cases of `lib.transpose`.

The new code implements only part of the functionality of[ `mkl_?imatcopy`](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2024-2/mkl-imatcopy.html). Conjugate transposes and non-square matrices are not supported.

~~Note: the source contains an `omp for collapse(2)` clause with a non-rectangular loop. Collapsing non-rectangular loops is [supported in OpenMP 5.0](https://developers.redhat.com/blog/2021/05/03/new-features-in-openmp-5-0-and-5-1#openmp_5_0_features) and works in GCC >= 11 and Clang >= 10. If this OpenMP feature is too new to use, let me know and I'll remove it.~~

I collapsed the loop nest manually to pass the CI tests.

### CODE REVIEW

File: pyscf/lib/np_helper/imatcopy.c [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+}
+
+
+static inline void dtranspose_scale_tile_offdiag(double *A, const int ii, const int jj,
+                                          int lda, const double alpha) {
+  for (int j = jj; j < jj + TILESIZE; j++) {
+#pragma omp simd
+    for(int i = ii; i < ii + TILESIZE; i++) {
+      const double tmp = ELEM_AT(A, i, j, lda);
+      ELEM_AT(A, i, j, lda) = alpha * ELEM_AT(A, j, i, lda);
```
Comment (sunqm) [inline pyscf/lib/np_helper/imatcopy.c:44]: 1. Remove the macro, simply using `A[i*lda+j]`
2. cast lda to size_t type

File: pyscf/lib/np_helper/imatcopy.c [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+/*
+ *    The following loop nest is equivalent to:
+ *    for(int iouter = 1; iouter < ntiles; iouter++)
+ *      for(int jouter = 0; jouter < iouter; jouter++)
+ *
+ *    See 10.1109/IPDPS.2017.34. 
+ */
+    int first_iteration = 1;
+    int iouter, jouter;
+#pragma omp for nowait
```
Comment (sunqm) [inline pyscf/lib/np_helper/imatcopy.c:153]: Set schedule(static) explicitly

File: pyscf/lib/np_helper/imatcopy.c [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+    for (int imat = 0; imat < nmat; imat++) {
+      NPomp_d_itranspose_scale(n, alpha, A + imat * matstride, lda);
+    }
+  }
+}
+
+void NPomp_ztensor_itranspose_scale021(const long long matstride, int nmat, int n, const double complex *alpha,
+                                      double complex *A, int lda)
+{
+    if (omp_get_num_threads() >= nmat) {
```
Comment (sunqm) [inline pyscf/lib/np_helper/imatcopy.c:364]: Please keep the code simple. The basic OpenMP parallelization is sufficient here. Not necessary to introduce the nested omp directives.

File: pyscf/lib/np_helper/imatcopy.c (hunk 1 above)
Comment (chillenb) [reply pyscf/lib/np_helper/imatcopy.c]: done!

File: pyscf/lib/np_helper/imatcopy.c (hunk 3 above)
Comment (chillenb) [reply pyscf/lib/np_helper/imatcopy.c]: OK!

**Review by sunqm — APPROVED** [review]

**Pull request merged**